kani-compiler/src/main.rs

```diff
@@ -31,152 +31,27 @@ extern crate tempfile;
-                Box::new(codegen_cprover_gotoc::GotocCodegenBackend::new(&Rc::new(queries)))
-            })));
-        } else {
-            return Err("Kani was configured without 'cprover' feature. You must enable this \
-            feature in order to use --goto-c argument.");
-        }
-    }
-    compiler.run().or(Err("Failed to compile crate."))
-}
-
-/// Empty struct since we don't support any callbacks yet.
-struct KaniCallbacks {}
-
-/// Use default function implementations.
-impl Callbacks for KaniCallbacks {
-    fn config(&mut self, config: &mut Config) {
-        if matches!(config.opts.error_format, ErrorOutputType::Json { .. }) {
-            json_panic_hook();
-        }
-    }
-}
-
-/// Generate the arguments to pass to rustc_driver.
-fn generate_rustc_args(args: &ArgMatches) -> Vec<String> {
-    let mut rustc_args = vec![String::from("rustc")];
-    if args.get_flag(parser::RUSTC_VERSION) {
-        rustc_args.push(String::from("--version"))
-    }
-
-    if args.get_flag(parser::JSON_OUTPUT) {
-        rustc_args.push(String::from("--error-format=json"));
-    }
-
-    if let Some(extra_flags) = args.get_raw(parser::RUSTC_OPTIONS) {
-        extra_flags.for_each(|arg| rustc_args.push(convert_arg(arg)));
-    }
-    tracing::debug!(?rustc_args, "Compile");
-    rustc_args
-}
-
-/// Convert an argument from OsStr to String.
-/// If conversion fails, panic with a custom message.
-fn convert_arg(arg: &OsStr) -> String {
-    arg.to_str().expect(format!("[Error] Cannot parse argument \"{arg:?}\".").as_str()).to_string()
-}
-
-/// Find the stub mapping for the given harness.
-///
-/// This function is necessary because Kani currently allows a harness to be
-/// specified by a partially qualified name, whereas stub mappings use fully
-/// qualified names.
-fn find_harness_stub_mapping(
-    harness: &str,
-    stub_mappings: FxHashMap<String, FxHashMap<DefPathHash, DefPathHash>>,
-) -> Option<FxHashMap<DefPathHash, DefPathHash>> {
-    let suffix = String::from("::") + harness;
-    for (name, mapping) in stub_mappings {
-        if name == harness || name.ends_with(&suffix) {
-            return Some(mapping);
-        }
-    }
-    None
-}
-
-#[cfg(test)]
-mod args_test {
-    use super::*;
-    use crate::parser;
-    #[cfg(unix)]
-    #[test]
-    #[should_panic]
-    fn test_invalid_arg_fails() {
-        use std::ffi::OsString;
-        use std::os::unix::ffi::OsStrExt;
-
-        // The value 0x80 is an invalid character.
-        let source = [0x68, 0x65, 0x6C, 0x6C, 0x80];
-        let os_str = OsStr::from_bytes(&source[..]);
-        assert_eq!(os_str.to_str(), None);
-
-        let matches = parser::parser().get_matches_from(vec![
-            OsString::from("kani-compiler").as_os_str(),
-            OsString::from("--sysroot").as_os_str(),
-            OsString::from("any").as_os_str(),
-            os_str,
-        ]);
-        generate_rustc_args(&matches);
+    if kani_compiler {
```

Comment:
Should this be an assert instead? Are there cases where we invoke the driver without `--kani-compiler`?

Reply:
Yes. To build proc_macro dependencies and build scripts. Cargo also invokes without ``--kani-compiler` to retrieve information about the compiler such as version.